crates/walrus-sui/src/client/retry_client.rs

```diff
@@ -1254,25 +1256,104 @@ impl From<tonic::Status> for CheckpointRpcError {
     }
 }
 
+/// A wrapper around a `RpcClient` that recreates the client if it has failed too many times.
+/// This is useful for clients that sometimes get stuck and never recover.
+pub struct FallibleRpcClient {
+    /// The RPC URL.
+    rpc_url: String,
+    /// The underlying RPC client.
+    client: Arc<RwLock<Arc<RpcClient>>>,
+    /// The instant of the last successful RPC call.
+    last_success: AtomicInstant,
+    /// The number of failures since the last successful RPC call.
+    num_failures: AtomicUsize,
+}
+
+impl FallibleRpcClient {
+    /// The time window for which failures are counted.
+    const FAILURE_WINDOW: Duration = Duration::from_secs(600);
+    /// The maximum number of failures allowed.
+    const MAX_FAILURES: usize = 100;
+
+    /// Creates a new `FallibleRpcClient`.
+    pub fn new(rpc_url: String) -> anyhow::Result<Self> {
+        let client = Arc::new(RpcClient::new(rpc_url.clone())?);
+        Ok(Self {
+            rpc_url,
+            client: Arc::new(RwLock::new(client)),
+            last_success: AtomicInstant::new(Instant::now()),
+            num_failures: AtomicUsize::new(0),
+        })
+    }
+
+    /// Returns the underlying RPC client.
+    pub async fn inner(&self) -> Arc<RpcClient> {
+        self.client.read().await.clone()
+    }
+
+    /// Calls the underlying RPC client with the given function.
+    pub async fn call<F, Fut, T>(&self, f: F, timeout: Duration) -> Result<T, tonic::Status>
+    where
+        F: FnOnce(Arc<RpcClient>) -> Fut,
+        Fut: Future<Output = Result<T, tonic::Status>>,
+    {
+        let client = self.client.read().await;
+        let result = tokio::time::timeout(timeout, f(client.clone())).await;
+
+        let result = match result {
+            Ok(Ok(result)) => Ok(result),
+            Ok(Err(status)) => Err(status),
+            Err(_) => Err(tonic::Status::deadline_exceeded("request timed out")),
+        };
+
+        let Ok(inner_result) = result else {
```

Comment:
Important: should we only start counting on certain error such as RPC error? For functioning error, e,g, the request objects/checkpoints do not exist, we probably shouldn't count them as failure that need client re-creation.

Reply:
I agree that would be ideal, but i think given we don't know how exactly the error manifests when the client gets stuck - it's probably safer for now to just do it for all errors. It's not going to be end of the world or performance impacting, given we re-create the client once every 10 minutes in some rare scenarios